### PR TITLE
fix(core): Add a new parameter `worker_event_max_payload` to kong.conf

### PR DESCRIPTION
**Pull request by liverpool8056**
### Summary

There is a limit for the size of payload allowed to be send in events lib. In some cases, Kong needs to send big payload. For example DAO:crud rely on events to invalidate cache. In these cases, the size of the payload may exceeds the limit due a large dao entity, this can prevents logics relying on the events from executing properly.

In this PR, a new parameter, `worker_events_max_payload` is added, which allows to specify the payload size the `worker_events` lib can accept. The default size is 64k, and the max allowed to set is 16M Bytes.

The corresponding PR for `worker_events` lib is [#37](https://github.com/Kong/lua-resty-events/pull/37)

### Checklist

- [x] The Pull Request has tests
- [x] There's an entry in the CHANGELOG
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Full changelog

* [Implement ...]

### Issue reference

FTI-4963

### CODE REVIEW

**Comment by chronolaw** [comment]
We should bumpe lua-resty-events to 0.2.0, see: #11203

File: kong/global.lua [hunk 1]
```diff
@@ -182,17 +182,21 @@ function _GLOBAL.init_worker_events()
                  and configuration.prefix
                  or require("pl.path").abspath(ngx.config.prefix())
 
+  local max_payload_len = configuration
```
Comment (chronolaw) [inline kong/global.lua:185]: should add entry in `conf_loader/init.lua`

File: kong.conf.default [hunk 2]
```diff
@@ -269,6 +269,11 @@
                                  # Similarly to `error_template_html`, the template
                                  # is required to contain one single `%s` placeholder for
                                  # the error message.
+
+#worker_events_max_payload =     # Specifies the max payload size the `worker_events` can
```
Comment (chronolaw) [inline kong.conf.default:273]: `worker_events_max_payload_size` recommanded.

Comment (liverpool8056) [reply kong.conf.default:273]: I gave the name referring to the parameter `cluster_max_payload` which is the existed one defining the size of a  payload, do you think if it's better to keep the shape of the line?

Comment (chronolaw) [reply kong.conf.default:273]: sound reasonable.

File: kong.conf.default [hunk 3]
```diff
@@ -269,6 +269,11 @@
...
+                                 # accept.
+                                 # Default is 64k - 1024 * 64. The max allowed is 16M Bytes.
```
Comment (dndx) [inline kong.conf.default:275]:
```suggestion
                                 # Default is 64KB. The max allowed is 16MB.
                                 # **Note:** increasing this value has potential
                                 # negative impact on Kong's response latency and memory usage.
```

File: kong/global.lua (hunk 1 above)
Comment (liverpool8056) [reply kong/global.lua:185]: done

File: CHANGELOG.md [hunk 4]
```diff
@@ -36,6 +36,10 @@
 
 #### Core
 
+- **worker_events**: Support specifying the payload size the worker_events can accept. 
+  Default is 64k, the max allowed size is 16M Bytes.
+  [#11214](https://github.com/Kong/kong/pull/11214)
```
Comment (chronolaw) [inline CHANGELOG.md:41]: Need a version bump entry in `Dependencies`

File: kong/templates/kong_defaults.lua [hunk 5]
```diff
@@ -42,6 +42,7 @@ cluster_dp_labels = NONE
 lmdb_environment_path = dbless.lmdb
 lmdb_map_size = 2048m
 mem_cache_size = 128m
+worker_events_max_payload = 65536
```
Comment (chronolaw) [inline kong/templates/kong_defaults.lua:45]: should be `65535`, which is the limitation before lua-resty-events `0.2.0`.

File: spec/02-integration/07-sdk/06-worker_events_spec.lua [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+          local status_code = allowed_size and 200 or 500
+          local msg = allowed_size and "ok" or "post " .. payload_type .. 
+            " failed, err: \"failed to publish event: payload exceeds the limitation (61440)\""
+          local body = assert.res_status(status_code, res)
+          assert.equal(body, msg)
+        end)
+      end
+    end)
+  end
+end)
```
Comment (chronolaw) [inline spec/02-integration/07-sdk/06-worker_events_spec.lua:156]: need a new line at the end of file.

File: CHANGELOG.md [hunk 7]
```diff
@@ -126,6 +130,8 @@
   [#11090](https://github.com/Kong/kong/pull/11090)
 - Remove kong branding from kong HTML error template.
   [#11150](https://github.com/Kong/kong/pull/11150)
+- `lua-resty-events` has been bumped to version 0.2.0.
```
Comment (chronolaw) [inline CHANGELOG.md:133]: should not be here. please refer to: https://github.com/Kong/kong/pull/11203/files

File: kong/conf_loader/init.lua [hunk 8]
```diff
@@ -321,6 +321,7 @@ local CONF_PARSERS = {
       replacement = "nginx_main_worker_processes",
     },
   },
+  worker_events_max_payload = { typ = "number" },
```
Comment (chronolaw) [inline kong/conf_loader/init.lua:324]: Add a blank line before this line.

File: CHANGELOG.md [hunk 9]
```diff
@@ -36,6 +36,10 @@
 
 #### Core
 
+- **worker_events**: Support specifying the payload size the worker_events can accept. 
+  Default is 64k, the max allowed size is 16M Bytes.
```
Comment (chronolaw) [inline CHANGELOG.md:40]:
```suggestion
  Default is 64KB, the max allowed size is 16MB.
```

File: spec/02-integration/07-sdk/06-worker_events_spec.lua [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+          location = /payload_string {
+            content_by_lua_block {
+              local PAYLOAD_TOO_BIG_ERR = "failed to publish event: payload too big"
+              local SOURCE, EVENT       = "foo", "string"
+              local worker_events       = kong.worker_events
+              local cjson               = require "cjson.safe"
+              local payload_received
+
+              local function generate_data()
+                return string.rep("X", %s)
```
Comment (chronolaw) [inline spec/02-integration/07-sdk/06-worker_events_spec.lua:26]: should be `%d`?

File: spec/02-integration/07-sdk/06-worker_events_spec.lua [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+  local fixtures = {
+    http_mock = {
+      worker_events = [[
+        server {
+          server_name example.com;
+          listen %s;
+
+          location = /payload_string {
+            content_by_lua_block {
+              local PAYLOAD_TOO_BIG_ERR = "failed to publish event: payload too big"
```
Comment (chronolaw) [inline spec/02-integration/07-sdk/06-worker_events_spec.lua:19]: `PAYLOAD_TOO_BIG_ERR` seems useless?

File: spec/02-integration/07-sdk/06-worker_events_spec.lua [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+
+              ngx.status = ngx.HTTP_OK
+              ngx.say("ok")
+              ngx.exit(200)
+            }
+          }
+
+          location = /payload_table {
+            content_by_lua_block {
+              local PAYLOAD_TOO_BIG_ERR       = "failed to publish event: payload too big"
```
Comment (chronolaw) [inline spec/02-integration/07-sdk/06-worker_events_spec.lua:65]: `PAYLOAD_TOO_BIG_ERR` seems useless?

File: spec/02-integration/07-sdk/06-worker_events_spec.lua (hunk 11 above)
Comment (liverpool8056) [reply spec/02-integration/07-sdk/06-worker_events_spec.lua]: done

File: spec/02-integration/07-sdk/06-worker_events_spec.lua (hunk 12 above)
Comment (liverpool8056) [reply spec/02-integration/07-sdk/06-worker_events_spec.lua]: done

File: spec/02-integration/07-sdk/06-worker_events_spec.lua (hunk 10 above)
Comment (liverpool8056) [reply spec/02-integration/07-sdk/06-worker_events_spec.lua:25]: it seems not matter here, we offer a string which actually presents an integer.

File: spec/02-integration/07-sdk/06-worker_events_spec.lua [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+        }
+      }
+    ]],
+  }
+}
+
+for _, max_payload in ipairs(max_payloads) do
+  local allowed_size = max_payload > payload_size
+  local less_or_greater = allowed_size and ">" or "<"
+  describe("when max_payload " .. less_or_greater .. " payload_size, ", function()
```
Comment (chronolaw) [inline spec/02-integration/07-sdk/06-worker_events_spec.lua:117]:
```suggestion

  describe("worker_events [when max_payload " .. less_or_greater .. " payload_size] ", function()
```

File: spec/02-integration/07-sdk/06-worker_events_spec.lua [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+            local ok, err = worker_events.register(function(data)
+              payload_received = data
+            end, SOURCE, EVENT)
+
+            -- when payload is a string
+            local PAYLOAD = generate_data()
+            local ok, err = worker_events.post(SOURCE, EVENT, PAYLOAD)
+            if not ok then
+              ngx.status = ngx.HTTP_INTERNAL_SERVER_ERROR
+              ngx.say("post string failed, err: " .. cjson.encode(err))
```
Comment (chronolaw) [inline spec/02-integration/07-sdk/06-worker_events_spec.lua:47]: `cjson.encode` seems unnecessary.

File: spec/02-integration/07-sdk/06-worker_events_spec.lua [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+
+            -- when payload is a table
+            PAYLOAD = {
+              foo = 'bar',
+              data = generate_data()
+            }
+            local ok, err = worker_events.post(SOURCE, EVENT, PAYLOAD)
+            if not ok then
+              ngx.status = ngx.HTTP_INTERNAL_SERVER_ERROR
+              ngx.say("post table failed, err: " .. cjson.encode(err))
```
Comment (chronolaw) [inline spec/02-integration/07-sdk/06-worker_events_spec.lua:96]: `cjson.encode` seems unnecessary. and remove `require "cjson.safe"`

File: spec/02-integration/07-sdk/06-worker_events_spec.lua [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+              return res
+            end
+
+            -- subscribe
+            local ok, err = worker_events.register(function(data)
+              payload_received = data
+            end, SOURCE, EVENT)
+
+            -- when payload is a table
+            PAYLOAD = {
```
Comment (chronolaw) [inline spec/02-integration/07-sdk/06-worker_events_spec.lua:89]:
```suggestion
            local PAYLOAD = {
```

**Review by chronolaw — APPROVED** [review]

**Review by Tieske — COMMENTED** [review]
what I get from the code is that it is specified in bytes. But that is not in the description, it says 64KB there. That's confusing.

Also since the numbers are never specified in byte-increments, maybe it should be specified in KB to begin with. But please check what other parameters of the same type we have, and ensure it is consistent with those.

**Review by Tieske — DISMISSED** [review]
see my comment

**Comment by Tieske** [comment]
also; afaik the config file is parsed and rendered on the docs website, so maybe @lena-larionova should also have a look

File: kong.conf.default [hunk 17]
```diff
@@ -269,6 +269,13 @@
                                  # Similarly to `error_template_html`, the template
                                  # is required to contain one single `%s` placeholder for
                                  # the error message.
+
+#worker_events_max_payload =     # Specifies the max payload size the `worker_events` can
+                                 # accept.
+                                 # Default is 65535. The max allowed is 16777215.
+                                 # **Note:** increasing this value has potential
+                                 # negative impact on Kong's response latency and memory usage.
```
Comment (lena-larionova) [inline kong.conf.default:277]:
```suggestion
#worker_events_max_payload = 65535  # Specifies the max payload size the 
                                    #`worker_events` can accept. The max allowed is 
                                    # 16777215.
                                    # **Note:** Increasing this value has potential
                                    # negative impact on Kong's response latency and memory usage.
```
Default value doesn't need to be in the description, it should be in the first line.

Comment (chronolaw) [reply kong.conf.default]: done. thanks.

File: kong.conf.default [hunk 18]
```diff
@@ -269,6 +269,14 @@
                                  # Similarly to `error_template_html`, the template
                                  # is required to contain one single `%s` placeholder for
                                  # the error message.
+
+#worker_events_max_payload = 65535  # Specifies the max payload size the
```
Comment (VicYP) [inline kong.conf.default:273]: Do we still need to document the change?
@hbagdi @dndx

Comment (chronolaw) [reply kong.conf.default]: Removed.

**Comment by kikito** [comment]
This was not merged in time for Feature freeze so I have removed it from the milestone. Please get special approval from a director-level person in order to merge this feature before 3.4 GA.

**Review by ADD-SP — APPROVED** [review]

**Pull request merged**